src/unit.h

```diff
@@ -6,25 +6,165 @@
+    : boost::units::base_unit<tof_base_unit, boost::units::time_dimension, 4> {
+};
+
+// Create the units system using the make_system utility
+typedef boost::units::make_system<wavelength_base_unit, counts_base_unit,
+                                  energy_base_unit, tof_base_unit>::type
+    units_system;
+
+typedef boost::units::unit<counts_dimension, units_system> counts;
+typedef boost::units::unit<boost::units::length_dimension, units_system>
+    wavelength;
+typedef boost::units::unit<boost::units::energy_dimension, units_system> energy;
+typedef boost::units::unit<boost::units::time_dimension, units_system> tof;
+
+/// unit constants
+BOOST_UNITS_STATIC_CONSTANT(lambda, wavelength);
+BOOST_UNITS_STATIC_CONSTANT(lambdas, wavelength);
+BOOST_UNITS_STATIC_CONSTANT(meV, energy);
+BOOST_UNITS_STATIC_CONSTANT(meVs, energy);
+BOOST_UNITS_STATIC_CONSTANT(microsecond, tof);
+BOOST_UNITS_STATIC_CONSTANT(microseconds, tof);
+
+} // namespace datasetunits
+
+// Define conversion factors: the conversion will work both ways with a single
+// macro
+
+// Convert angstroms to SI meters
+BOOST_UNITS_DEFINE_CONVERSION_FACTOR(datasetunits::wavelength_base_unit,
+                                     boost::units::si::length, double, 1.0e-10);
+// Convert meV to SI Joule
+BOOST_UNITS_DEFINE_CONVERSION_FACTOR(
+    datasetunits::energy_base_unit, boost::units::si::energy, double,
+    1.0e-3 * boost::units::si::constants::codata::e.value().value());
+// Convert tof microseconds to SI seconds
+BOOST_UNITS_DEFINE_CONVERSION_FACTOR(datasetunits::tof_base_unit,
+                                     boost::units::si::time, double, 1.0e-6);
+
+template <>
+struct boost::units::base_unit_info<datasetunits::wavelength_base_unit> {
+  static std::string name() { return "angstroms"; }
+  static std::string symbol() { return "AA"; }
+};
+
+template <>
+struct boost::units::base_unit_info<datasetunits::counts_base_unit> {
+  static std::string name() { return "counts"; }
+};
+
+template <>
+struct boost::units::base_unit_info<datasetunits::energy_base_unit> {
+  static std::string name() { return "milli-electronvolt"; }
+  static std::string symbol() { return "meV"; }
+};
+
+template <> struct boost::units::base_unit_info<datasetunits::tof_base_unit> {
+  static std::string name() { return "microseconds"; }
+  static std::string symbol() { return "us"; }
+};
+
 class Unit {
 public:
+  // Define a std::variant which will hold the set of allowed units.
+  // Any unit that does not exist in the variant will either fail to compile or
+  // throw a std::runtime_error during operations such as multiplication or
+  // division.
+  // 
+  // TODO: maybe it is possible to create a helper that will automatically
+  // generate the squares for variance?
+  // The following was attempted but did not succeed:
+  //  template <class... Ts> struct unit_and_variance {
+  //  using type =
+  //    std::variant<Ts..., decltype(std::declval<Ts>()*std::declval<Ts>())...>;
+  //  };
```

Comment:
Had thought about this again, I think you can do it like this, which is almost the same convenience:

1. list all units without powers (only power 1).
2. Use `unit_and_variance` to also get the squares (gives you all powers of 2).
3. Manually add any higher powers you may need (e.g., m^3, m^4, m^6).

Reply:
I understand steps 1 and 2, but i'm not sure how I would go about extending the set manually after using `unit_and_variance`...

Reply:
One way to do it is to pass two parameter packs to `unit_and_variance`:

```cpp
namespace detail {
template <class... Ts, class... Extra>
std::variant<Ts..., decltype(std::declval<Ts>()*std::declval<Ts>())..., Extra...>
make_unit(const std::tuple<Ts...> &, const std::tuple<Extra...> &) { return {}; }
}

using unit_t = decltype(detail::make_unit(std::make_tuple(none, m, s, kg, counts), std::make_tuple(m4, m6)));
```

Note that you cannot pass 2 parameter packs directly. Therefore you need to pass 2 tuples instead and deduce the packs (the advantage seems to be that you do not need the `declval` anymore when calling the helper).

You might need to throw in something like `std::remove_reference_t` in some places, not sure.